### PR TITLE
libEnsemble: add v0.10.0

### PR DESCRIPTION
**Pull request by shuds13**

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-libensemble/package.py [hunk 1]
```diff
@@ -45,6 +46,7 @@ class PyLibensemble(PythonPackage):
     variant("deap", default=False, description="Install with DEAP")
     variant("tasmanian", default=False, description="Install with tasmanian")
     variant("pyyaml", default=False, description="Install with pyyaml")
+    variant("pytoml", default=False, description="Install with pytoml")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-libensemble/package.py:48]: All of the following dependencies are required and shouldn't have variants, at least in the latest version:
```python
    install_requires=["numpy", "psutil", "setuptools", "pydantic", "tomli", "pyyaml"],   
```
The only variants I see listed explicitly in the package are +extras and +docs.

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-libensemble/package.py (hunk 1 above)
Comment (shuds13) [reply var/spack/repos/builtin/packages/py-libensemble/package.py:48]: Fixed

File: var/spack/repos/builtin/packages/py-libensemble/package.py [hunk 2]
```diff
@@ -44,7 +45,6 @@ class PyLibensemble(PythonPackage):
     variant("mpmath", default=False, description="Install with mpmath")
     variant("deap", default=False, description="Install with DEAP")
     variant("tasmanian", default=False, description="Install with tasmanian")
-    variant("pyyaml", default=False, description="Install with pyyaml")
 
     depends_on("py-setuptools", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-libensemble/package.py]:
```suggestion
    depends_on("py-setuptools", type=("build", "run"))
```
It's listed under install_requires

File: var/spack/repos/builtin/packages/py-libensemble/package.py [hunk 3]
```diff
@@ -44,7 +45,6 @@ class PyLibensemble(PythonPackage):
     variant("mpmath", default=False, description="Install with mpmath")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-libensemble/package.py:45]: Let's create a single `+extras` variant to control these. This matches the pip installation method.

File: var/spack/repos/builtin/packages/py-libensemble/package.py [hunk 4]
```diff
@@ -60,7 +60,8 @@ class PyLibensemble(PythonPackage):
     depends_on("py-mpmath", type=("build", "run"), when="+mpmath")
     depends_on("py-deap", type=("build", "run"), when="+deap")
     depends_on("tasmanian+python", type=("build", "run"), when="+tasmanian")
-    depends_on("py-pyyaml", type=("build", "run"), when="+pyyaml")
+    depends_on("py-pyyaml", type=("build", "run"), when="@:0.10.0")
+    depends_on("py-toml", type=("build", "run"), when="@:0.10.0")
     conflicts("~mpi", when="@:0.4.1")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-libensemble/package.py:66]: Missing dependency on py-pydantic

File: var/spack/repos/builtin/packages/py-libensemble/package.py [hunk 5]
```diff
@@ -60,7 +60,8 @@ class PyLibensemble(PythonPackage):
     depends_on("py-mpmath", type=("build", "run"), when="+mpmath")
     depends_on("py-deap", type=("build", "run"), when="+deap")
     depends_on("tasmanian+python", type=("build", "run"), when="+tasmanian")
-    depends_on("py-pyyaml", type=("build", "run"), when="+pyyaml")
+    depends_on("py-pyyaml", type=("build", "run"), when="@:0.10.0")
+    depends_on("py-toml", type=("build", "run"), when="@:0.10.0")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-libensemble/package.py]:
```suggestion
    depends_on("py-pyyaml", type=("build", "run"), when="@0.10:")
    depends_on("py-pyyaml", type=("build", "run"), when="@:0.9 +extras")
    depends_on("py-tomli", type=("build", "run"), when="@0.10:")
```
Also required in develop

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-libensemble/package.py (hunk 3 above)
Comment (shuds13) [reply var/spack/repos/builtin/packages/py-libensemble/package.py:45]: Well, the "extras" are to run all the tests, which use various packages. I don't think its likely one person would need all of these things.

File: var/spack/repos/builtin/packages/py-libensemble/package.py (hunk 2 above)
Comment (shuds13) [reply var/spack/repos/builtin/packages/py-libensemble/package.py]: I think this is only needed for installation.

**Comment by shuds13** [comment]
@jlnav Can you take over this, when you are back?

File: var/spack/repos/builtin/packages/py-libensemble/package.py (hunk 2 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-libensemble/package.py]: You're right that it isn't needed after install. But it's listed under `install_requires`, so `pip check` will complain if it isn't found. The correct solution is to move it to `setup_requires`. But for now, it's a build/run dep.

**Comment by jlnav** [comment]
> 

@shuds13 , Sure thing, I've made changes and can open a PR into your fork; which should also update this PR (?). I think you need to manually allow that capability yourself though.

File: var/spack/repos/builtin/packages/py-libensemble/package.py (hunk 2 above)
Comment (shuds13) [reply var/spack/repos/builtin/packages/py-libensemble/package.py]: Added

File: var/spack/repos/builtin/packages/py-libensemble/package.py (hunk 4 above)
Comment (shuds13) [reply var/spack/repos/builtin/packages/py-libensemble/package.py:66]: Added

**Comment by shuds13** [comment]
@adamjstewart I think this is how we want it now. @jlnav you can close the other PR.

File: var/spack/repos/builtin/packages/py-libensemble/package.py [hunk 6]
```diff
@@ -60,7 +60,9 @@ class PyLibensemble(PythonPackage):
     depends_on("py-mpmath", type=("build", "run"), when="+mpmath")
     depends_on("py-deap", type=("build", "run"), when="+deap")
     depends_on("tasmanian+python", type=("build", "run"), when="+tasmanian")
-    depends_on("py-pyyaml", type=("build", "run"), when="+pyyaml")
+    depends_on("py-pyyaml", type=("build", "run"), when="@:0.10.0")
+    depends_on("py-tomli", type=("build", "run"), when="@:0.10.0")
+    depends_on("py-pydantic", type=("build", "run"), when="@:0.10.0")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-libensemble/package.py]:
```suggestion
    depends_on("py-pyyaml", type=("build", "run"), when="@0.10:")
    depends_on("py-tomli", type=("build", "run"), when="@0.10:")
    depends_on("py-pydantic", type=("build", "run"), when="@0.10:")
```

**Review by adamjstewart — COMMENTED** [review]
Not required, but I usually like to list deps in the same order they are listed in `setup.py`. Makes it easier for me to review.

**Comment by shuds13** [comment]
@adamjstewart Is this ready to pull in now?

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**